# Notebook 01 – Data Ingestion

## Objective
Extract and preprocess YouTube transcripts for use in a Retrieval-Augmented Generation (RAG) pipeline.

## Input
YouTube Video ID

## Output
Clean transcript saved as .txt file

## Methodology
- Fetch transcript using youtube-transcript-api
- Convert structured segments into continuous text
- Save for downstream processing


In [1]:
"""
PROJECT: 
NeuralTranscript: A RAG-Based Semantic Search & Q&A System for YouTube Content

-------------------------------------------------------------------------
AUTHOR: Engr. Inam Ullah Khan
Master's Student in Data Science | Al-Farabi Kazakh National University
-------------------------------------------------------------------------
"""

import os
from typing import List, Any
from youtube_transcript_api import YouTubeTranscriptApi

# --- 1. CONFIGURATION ---
# Video: Demis Hassabis (DeepMind CEO) on AI
VIDEO_ID = "Gfr50f6ZBvo" 
OUTPUT_FOLDER = "data/transcripts"

# --- 2. CORE FUNCTIONS ---

def fetch_youtube_transcript(video_id: str) -> List[Any]:
    """
    Retrieves the raw transcript data using the latest API standards.
    
    Args:
        video_id (str): The unique 11-character YouTube video ID.
        
    Returns:
        List[Any]: Raw transcript segments containing text and timestamps.
    """
    print(f"📡 Accessing YouTube API for Video ID: {video_id}...")
    try:
        # Initializing the API instance (v1.x.x compatibility)
        api_instance = YouTubeTranscriptApi()
        
        # Fetching transcript and converting to standard list format
        raw_data = api_instance.fetch(video_id).to_raw_data()
        return raw_data
    
    except Exception as e:
        print(f"❌ Error: Could not retrieve transcript. Details: {e}")
        return []

def process_transcript_to_text(raw_transcript: List[Any]) -> str:
    """
    Cleans and unifies transcript segments into a single cohesive string.
    
    Args:
        raw_transcript (List[Any]): List of transcript dictionaries.
        
    Returns:
        str: A cleaned block of text ready for NLP tasks.
    """
    print("🧹 Cleaning and unifying transcript text...")
    
    # Extract 'text' field while handling potential object/dict variability
    text_segments = []
    for segment in raw_transcript:
        if isinstance(segment, dict):
            text_segments.append(segment.get("text", ""))
        else:
            text_segments.append(getattr(segment, 'text', ""))
            
    # Join segments and remove excess whitespace for cleaner processing
    unified_text = " ".join(text_segments).replace("  ", " ").strip()
    return unified_text

def save_to_disk(text_content: str, filename: str):
    """
    Persists the cleaned text to a local file for downstream processing.
    """
    if not os.path.exists(OUTPUT_FOLDER):
        os.makedirs(OUTPUT_FOLDER)
        print(f"📁 Created directory: {OUTPUT_FOLDER}")
        
    file_path = os.path.join(OUTPUT_FOLDER, f"{filename}.txt")
    
    with open(file_path, "w", encoding="utf-8") as f:
        f.write(text_content)
    
    print(f"💾 Success! Cleaned transcript saved to: {file_path}")

# --- 3. MAIN PIPELINE EXECUTION ---

if __name__ == "__main__":
    print(f"--- Starting NeuralTranscript Ingestion ---")
    
    # Step 1: Extraction
    raw_transcript_data = fetch_youtube_transcript(VIDEO_ID)
    
    if raw_transcript_data:
        # Step 2: Transformation
        final_text = process_transcript_to_text(raw_transcript_data)
        
        if final_text:
            print(f"📊 Stats: {len(final_text)} characters processed.")
            
            # Step 3: Loading (Saving)
            save_to_disk(final_text, VIDEO_ID)
            
            # Final Preview
            print(f"\n📝 PREVIEW (First 250 chars):\n{final_text[:250]}...")
        else:
            print("⚠️ Warning: Transcript was empty after processing.")
    else:
        print("❌ Pipeline halted: No data to process.")

--- Starting NeuralTranscript Ingestion ---
📡 Accessing YouTube API for Video ID: Gfr50f6ZBvo...
🧹 Cleaning and unifying transcript text...
📊 Stats: 133836 characters processed.
💾 Success! Cleaned transcript saved to: data/transcripts\Gfr50f6ZBvo.txt

📝 PREVIEW (First 250 chars):
the following is a conversation with demus hasabis ceo and co-founder of deepmind a company that has published and builds some of the most incredible artificial intelligence systems in the history of computing including alfred zero that learned all b...


## Observations

- Transcript successfully retrieved without errors.
- Total character count: 133836.
- Transcript saved locally for downstream chunking.
- No missing segments detected.


## Summary

The data ingestion stage successfully extracted and cleaned the YouTube transcript.
The resulting text file will serve as the raw document for semantic chunking and embedding generation in subsequent stages of the RAG pipeline.
This completes the first stage of the system architecture.

**Next step:** Text chunking and preprocessing (`02_chunking_analysis.ipynb`)
